### PR TITLE
feat: add resourceDiscovery status condition

### PR DESCRIPTION
**Pull request by jan--f**
feat: add resourceDiscovery status condition

    The ResourceSelector field in a MonitoringStack CR can be set to nil.
    This is passed on to the Prometheus CR and can be used to disable all
    service discovery and thus most scraping (we add prometheus and
    alertmanager scrape jobs via additionalScrapeConfigs).
    When this is configured we now set a new status condition to alert users
    to the fact that no service discovery is configured.

Fixes: https://issues.redhat.com/browse/MON-2864

Signed-off-by: Jan Fajerski <jfajersk@redhat.com>

### CODE REVIEW

**Review by JoaoBraveCoding — DISMISSED** [review]
/lgtm

**Review by sthaha — COMMENTED** [review]
Nice! lgtm

I do feel that we need a test that asserts that explicit `nil` is preserved and not specifying anything sets `{}` 

To actually cover the usecase we discussed, I think we need a e2e test that 
* sets resourceSelector
* resets to nil
* validates selectors in Prometheus reflect the same 
* 

WDYT @simonpasquier ?

**Comment by jan--f** [comment]
i'm not sure OBO is the right place for these tests. Prometheus-operator would be a better place for this, wouldn't it?
[A test for `ServiceMonitorSelector = nil`](https://github.com/prometheus-operator/prometheus-operator/blob/66f07d0267e224391d16e0ff20d771ddd0b61404/test/e2e/prometheus_test.go#L809-L822) already exists in the prometheus-operator repo. If a check for "No generated scrape jobs in Prometheus" makes sense in an e2e test, I'd argue over there is the right place.

**Comment by sthaha** [comment]
> * sets resourceSelector
> * resets to nil
> * validates selectors in Prometheus reflect the same



> i'm not sure OBO is the right place for these tests. Prometheus-operator would be a better place for this, wouldn't it? [A test for `ServiceMonitorSelector = nil`](https://github.com/prometheus-operator/prometheus-operator/blob/66f07d0267e224391d16e0ff20d771ddd0b61404/test/e2e/prometheus_test.go#L809-L822) already exists in the prometheus-operator repo.

I am not sure I agree ... The suggestion was to only validate the functionality provided by ObO .. i.e. setting `resourceSelector` of the **MonitoringStack** from `{}` (default) to `nil` does indeed update/generate `Prometheus` (which ObO generates) that has the `<resource>Selector` set to nil. 
We are not validating the functionality of the PO rather only validating the functionality provided by Obo. 


 > If a check for "No generated scrape jobs in Prometheus" makes sense in an e2e test, I'd argue over there is the right place.
This I agree is beyond the scope of the ObO tests but that isn't what I was requesting.

**Comment by jan--f** [comment]
> The suggestion was to only validate the functionality provided by ObO .. i.e. setting resourceSelector of the MonitoringStack from {} (default) to nil does indeed update/generate Prometheus (which ObO generates) that has the <resource>Selector set to nil.

Ahh I see, yeah thats indeed worth testing here. I'll add that.

**Review by simonpasquier — COMMENTED** [review]
late to this but yes I agree with @sthaha that end-to-end tests for the following are valuable:
* By default, ObO provisions underlying objects with the empty selector.
* `spec.resourceSelector: nil` is supported and propagates to the underlying objects.

**Comment by jan--f** [comment]
Update: I have a test for this but I'm seeing some unexpected behavior which I'd like to look into.
It seems like `+kubebuilder:default={}` on `ResourceSelector` doesn't have the desired effect and when a stack with a `null` selector is created, thanos side car doesn't start.

**Comment by sthaha** [comment]
> It seems like +kubebuilder:default={} on ResourceSelector doesn't have the desired effect and when a stack with a null selector is created,

I suspected `nil` will be set back to `{}` but this is even weirder :(

File: pkg/apis/monitoring/v1alpha1/types.go [hunk 1]
```diff
@@ -55,8 +55,10 @@ type MonitoringStackSpec struct {
 	LogLevel LogLevel `json:"logLevel,omitempty"`
 
 	// Label selector for Monitoring Stack Resources.
-	// +optional
-	ResourceSelector *metav1.LabelSelector `json:"resourceSelector,omitempty"`
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default={}
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/types.go:59]: `+kubebuilder:default` doesn't seem to play well with pointer fields: the default value doesn't appear in the generated YAML manifest of the CRD :thinking:

Comment (simonpasquier) [reply pkg/apis/monitoring/v1alpha1/types.go:59]: I suspect that kubebuilder's defaulting works well with value fields but not so much with pointers.

**Review by JoaoBraveCoding — COMMENTED** [review]
In general looks good! Just two comments but nothing blocking 👍

File: pkg/controllers/monitoring/monitoring-stack/conditions.go [hunk 2]
```diff
@@ -15,13 +15,17 @@ const (
...
+	ResourceSelectorIsNil          = "ResourceSelectorNil"
 	CannotReadPrometheusConditions = "Cannot read Prometheus status conditions"
 	AvailableMessage               = "Monitoring Stack is available"
 	SuccessfullyReconciledMessage  = "Monitoring Stack is successfully reconciled"
+	ResourceSelectorIsNilMessage   = "No resources will be discovered, ResourceSelector is nil"
+	ResourceDiscoveryOnMessage     = "Resource discovery is operational"
 	NoReason                       = "None"
 )
 
-func updateConditions(msConditions []v1alpha1.Condition, prom monv1.Prometheus, generation int64, recError error) []v1alpha1.Condition {
```
[the remark this answers is not in the document]
Comment (jan--f) [reply pkg/controllers/monitoring/monitoring-stack/conditions.go:24]: This is unfortunately necessary since the resource discovery condition is set based on a field of `MonitoringStack`. 
See https://github.com/rhobs/observability-operator/pull/223/files#diff-3c84e774a1e9ef2eafc525a973f4b4a273410b6b242fabda42e514c524328a38R64
And since the conditions are part of that anyway I simply changed the argument, instead of passing both the conditions list and the monitoring stack.

File: pkg/controllers/monitoring/monitoring-stack/conditions.go [hunk 3]
```diff
@@ -54,12 +60,44 @@ func updateConditions(msConditions []v1alpha1.Condition, prom monv1.Prometheus,
...
+			}
+			updatedConditions = append(updatedConditions, resourceDiscovery)
 		}
 	}
 
 	return updatedConditions
 }
 
+// updateResouceDiscovery updates the ResouceDiscoveryCondition based on the
+// ResrouceSelector in the MonitorinStack spec. A ResourceSelector of nil causes
```
Comment (tremes) [inline pkg/controllers/monitoring/monitoring-stack/conditions.go:76]: Nit: typo in `ResrouceSelector`

Comment (jan--f) [reply pkg/controllers/monitoring/monitoring-stack/conditions.go]: thanks, nice catch. For some reason typing `resource` is the bane of my existence.

File: pkg/controllers/monitoring/monitoring-stack/conditions.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (JoaoBraveCoding) [reply pkg/controllers/monitoring/monitoring-stack/conditions.go:24]: Ahh yes, missed that detail!

File: pkg/controllers/monitoring/monitoring-stack/conditions_test.go [hunk 4]
```diff
@@ -13,152 +13,220 @@ func TestUpdateConditions(t *testing.T) {
 	transitionTime := metav1.Now()
 
 	tt := []struct {
-		name                 string
-		originalMSConditions []v1alpha1.Condition
-		prometheus           monv1.Prometheus
-		generation           int64
-		recError             error
-		expectedResults      []v1alpha1.Condition
+		name             string
```
[the remark this answers is not in the document]
Comment (JoaoBraveCoding) [reply pkg/controllers/monitoring/monitoring-stack/conditions_test.go:16]: OK after that explanation and with the argument change, I think it makes sense to split!

**Review by JoaoBraveCoding — APPROVED** [review]
lgtm

File: pkg/apis/monitoring/v1alpha1/types.go [hunk 5]
```diff
@@ -58,8 +58,11 @@ type MonitoringStackSpec struct {
 	LogLevel LogLevel `json:"logLevel,omitempty"`
 
 	// Label selector for Monitoring Stack Resources.
+	// Set to the empty LabelSelector to monitoring everything.
```
Comment (sthaha) [inline pkg/apis/monitoring/v1alpha1/types.go:61]:
```suggestion
	// Set to an empty LabelSelector ({}) to monitoring everything.
```

File: pkg/controllers/monitoring/monitoring-stack/conditions.go [hunk 6]
```diff
@@ -15,13 +15,17 @@ const (
...
 	AvailableMessage               = "Monitoring Stack is available"
 	SuccessfullyReconciledMessage  = "Monitoring Stack is successfully reconciled"
+	ResourceSelectorIsNilMessage   = "No resources will be discovered, ResourceSelector is nil"
+	ResourceDiscoveryOnMessage     = "Resource discovery is operational"
 	NoReason                       = "None"
 )
 
-func updateConditions(msConditions []v1alpha1.Condition, prom monv1.Prometheus, generation int64, recError error) []v1alpha1.Condition {
+func updateConditions(ms *v1alpha1.MonitoringStack, prom monv1.Prometheus, recError error) []v1alpha1.Condition {
+	msConditions := ms.Status.Conditions
```
Comment (sthaha) [inline pkg/controllers/monitoring/monitoring-stack/conditions.go:28]: nit: could we move this close to where it is used? and perhaps we may not even need this.

File: pkg/controllers/monitoring/monitoring-stack/conditions.go [hunk 7]
```diff
@@ -54,12 +60,44 @@ func updateConditions(msConditions []v1alpha1.Condition, prom monv1.Prometheus,
 				reconciled.LastTransitionTime = metav1.Now()
 			}
 			updatedConditions = append(updatedConditions, reconciled)
+		case v1alpha1.ResourceDiscoveryCondition:
```
Comment (sthaha) [inline pkg/controllers/monitoring/monitoring-stack/conditions.go:63]: doesn't this mean, that a `ms` that is already deployed prior to this feature  will  never  have this condition set? Instead shouldn't we set all 3 conditions irrespective of them them having the condition?

File: test/e2e/monitoring_stack_controller_test.go [hunk 8]
```diff
@@ -122,6 +128,23 @@ func emptyStackCreatesPrometheus(t *testing.T) {
...
+	f.AssertResourceEventuallyExists(ms.Name, ms.Namespace, &monv1.Prometheus{})(t)
+
+	updatedMS := &stack.MonitoringStack{}
+	f.GetResourceWithRetry(t, ms.Name, ms.Namespace, updatedMS)
+	updatedMS.Spec.ResourceSelector = nil
+	err = f.K8sClient.Update(context.Background(), updatedMS)
+	assert.NilError(t, err, "failed to patch monitoring stack with nil resource selector")
+
+	prometheus := monv1.Prometheus{}
+	f.GetResourceWithRetry(t, updatedMS.Name, updatedMS.Namespace, &prometheus)
```
Comment (sthaha) [inline test/e2e/monitoring_stack_controller_test.go:144]: this may yield failure depending on the timing because `GetResourceWithRetry` may give us the old Prometheus ... we need to assert the generation has changed whenever we modify an exisiting stack.

File: pkg/apis/monitoring/v1alpha1/types.go [hunk 9]
```diff
@@ -58,8 +58,11 @@ type MonitoringStackSpec struct {
...
+	// Set to null to disable service discovery.
 	// +optional
-	ResourceSelector *metav1.LabelSelector `json:"resourceSelector,omitempty"`
+	// +nullable
```
Comment (sthaha) [inline pkg/apis/monitoring/v1alpha1/types.go:64]: Shouldn't we set the default to have service discovery enabled? i.e. - `{}` ?

File: test/e2e/monitoring_stack_controller_test.go [hunk 10]
```diff
@@ -626,10 +649,17 @@ func msNamespaceSelector(labels map[string]string) stackModifier {
...
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: stack.GroupVersion.String(),
+			Kind:       "MonitoringStack",
+		},
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      name,
 			Namespace: e2eTestNamespace,
 		},
+		Spec: stack.MonitoringStackSpec{
+			ResourceSelector: &metav1.LabelSelector{},
```
Comment (sthaha) [inline test/e2e/monitoring_stack_controller_test.go:661]: Isn't it better to use a `stackModifier` to explicitly set the ResourceSelector since the default is nil?

Comment (jan--f) [reply test/e2e/monitoring_stack_controller_test.go:661]: Given that the majority of tests expects the empty selector and this is a convenience function I went with this.
Causes less code changes to existing code too.

File: pkg/controllers/monitoring/monitoring-stack/conditions.go (hunk 6 above)
Comment (jan--f) [reply pkg/controllers/monitoring/monitoring-stack/conditions.go:28]: Its used in the very next line.

File: pkg/apis/monitoring/v1alpha1/types.go (hunk 5 above)
Comment (jan--f) [reply pkg/apis/monitoring/v1alpha1/types.go:61]: But isn't there just the one empty LabelSelector? Put :+1: on the `({})` bit.

File: pkg/controllers/monitoring/monitoring-stack/conditions.go (hunk 7 above)
Comment (jan--f) [reply pkg/controllers/monitoring/monitoring-stack/conditions.go:63]: Hm yeah good point.

File: pkg/apis/monitoring/v1alpha1/types.go (hunk 9 above)
Comment (jan--f) [reply pkg/apis/monitoring/v1alpha1/types.go:64]: Setting a default annotation here has no effect on the generated manifest. I haven't looked into the depth of the OpenAPI code but I suspect `default` and `nullable` doesn't go well together.

**Review by sthaha — APPROVED** [review]
lgtm ♥️
